gsy_framework/community_datasheet/community_datasheet_parser.py

```diff
@@ -0,0 +1,182 @@
+
+    @staticmethod
+    def _validate_grid(grid):
+        try:
+            scenario_validator(grid)
+        except ValidationError as ex:
+            raise CommunityDatasheetException(ex) from ex
+
+    def _get_assets_by_member(self) -> Dict:
+        """Return a mapping between each member and their assets."""
+        assets_by_member = {member_name: [] for member_name in self._datasheet.members}
+        asset_items = [
+            assets.items()
+            for assets in [self._datasheet.loads, self._datasheet.pvs, self._datasheet.storages]
+        ]
+        for member_name, assets in chain.from_iterable(asset_items):
+            assets_by_member[member_name].extend(assets)
+
+        if any((missing := member) not in self._datasheet.members for member in assets_by_member):
+            raise CommunityDatasheetException(
+                f'Member "{missing}" was defined in one of the asset sheets'  # noqa: F821
+                'but not in the "Community Members" sheet.')
+
+        return assets_by_member
+
+    @staticmethod
+    def _merge_profiles_into_assets(profiles: Dict, assets_by_member: Dict) -> None:
+        """Merge (in-place) each energy profile into the representation of its own asset."""
+        for asset in chain.from_iterable(assets_by_member.values()):
+            asset_name = asset["name"]
+            if asset_name not in profiles:
+                continue
+            asset_type = asset["type"]
+            try:
+                profile_key = PROFILE_KEYS_BY_TYPE[asset_type]
+            except KeyError as ex:
+                raise CommunityDatasheetException(
+                    f'Asset type "{asset_type}" is not supported.') from ex
+            asset[profile_key] = profiles[asset_name]
+
+    def _create_grid(self, assets_by_member: Dict) -> Dict:
+        members_grid = []
+        for member_name, assets in assets_by_member.items():
+            member_representation = self._create_member_representation(member_name)
+            member_representation["children"] = assets
+            members_grid.append(member_representation)
+
+        return {
+            "name": "Grid Market",
+            "tags": None,
+            "type": "Area",
+            "uuid": str(uuid.uuid4()),
+            "children": members_grid
+        }
+
+    def _create_member_representation(self, member_name: str) -> Dict:
+        return {
+            **self._datasheet.members[member_name],
+            **{
+                "name": member_name,
+                "tags": ["Home"],
+                "type": "Area",
+                "uuid": str(uuid.uuid4()),
+                "children": []
+            }}
+
+
+def parse_community_datasheet(filename):
+    """Parse the content of the community datasheet and return a JSON representation."""
+    parser = CommunityDatasheetParser(filename)
+    print(json.dumps(parser.parse(), indent=4))
+    # NOTE: validate datasheet before or after parsing?
```

Comment:
In general, when accepting user input, I'd always do some validation. Not only because users might add malicious input (not sure it's our case here) but mostly because - if something fails - the user needs to know what exactly was the issue so that they can fix it. And not all errors can be understood by the users (e.g. type errors or key errors wouldn't make much sense for them if we don't state what caused them).

If the validation only happens on gsy-web, the user won't get any meaningful information (because the errors they will see won't mention the datasheet fields, but only the grid fields (and the user doesn't know anything about them).

Also, the parser won't fail in some cases. The datasheet might be "formally" valid (e.g. it won't fail while we read the values) but it might contain reciprocally contradictory values. Or it might contain sets of values that are required together (like the Rebase ones). Or it wouldn't fail if we don't fill some mandatory field or if we fill them with inconsistent values.

This is also related to what we discussed in the last retrospective: we'd like to create hackathon tickets to implement (at least) 2 schema validations. One for grids received as input by gsy-web, the other for grids received as input by gsy-e. This might make the validation of the datasheet a bit less necessary (but still useful for the user IMO).